### DISCUSSION TITLE
Daly BMS integration

### DISCUSSION THREAD

**Discussion by Coolj1** [hoylabs/OpenDTU-OnBattery#100] [Ideas]
is it possible, to integrate a Daly-BMS?

**Comment by helgeerbe** [comment]
I don't know this BMS, but a quick search on google shows that there is an esphome project (https://esphome.io/components/sensor/daly_bms.html). So I guess with a serial UART interface (cable not bluetooth!) it should be possible.

**Reply by Coolj1** [reply]
i want to build a zero injektion solar system with a victron mppt 100/20, a Hoymiles hm-300, a VW-ID3 Battery, a Daly-BMS and a Shelly 3em with a standalone regulation, no Homeautomation like hopenhab... should regulate this. i think, one ESP32 can handle this, but my programming knowledge for arduino is very bad.

**Reply by Coolj1** [reply]
ich denke, es ist einfacher ich schreibe auf deutsch weiter😁,

es gibt für den Esp32 einen Basicinterpreter (Annex RDS Basic), damit kann ich die API-Abfrage für den Shelly und den OpenDTU sehr einfach machen (ich bin Elektroniker und kein Programmierer😢).

Das funktioniert auch ohne Probleme, jedoch beim setzen des Limits habe ich Probleme.
Mit meinem Raspi funktioniert das Beispiel von dir einwandfrei, jedoch kann ich den Syntax nicht direkt übernehmen.
hier mein Versuch:

message$="'data={"
message$=message$+chr$(34)+"serial"+chr$(34)+":"+chr$(34)+"112183822934"+chr$(34)
message$=message$+", "+chr$(34)+"limit_type"+chr$(34)+":"+"1"
message$=message$+", "+chr$(34)+"limit_value"+chr$(34)+":"+"50"
message$=message$+"}'"

wlog message$

a$ = WPOST$("http://admin:admin0815@192.168.1.228/api/limit/config", message$)
'a$ = WGET$("http://admin:admin0815@192.168.1.228/api/ntp/config")
wlog a$


mit dem Befehl WGET$ funktioniert es super... was mache ich falsch?

**Reply by helgeerbe** [reply]
Hast du es mal ohne die Single Quotes (') am Anfang und Ende versucht?

**Comment by Coolj1** [comment]
hi,danke für die Antwort, ja das hab ich. (siehe Screenshot)das Ergebnis sieht man im rechten Fenster, der findet die Variaben nicht.das was ich im Programm auskommentiert habe, habe ich versucht.das WGet funktioniert, aber die Datenübergabe mit dem WPost nicht😢kann man da irgend wie mitloggen, was da beim Open-DTU ankommt? 
![image](https://user-images.githubusercontent.com/125164924/218875642-3f31353b-394c-48dc-9488-d7aa92bd6921.png)

**Reply by helgeerbe** [reply]
According to RFC4627, the official media type for JavaScript Object Notation (JSON) objects is "application/json". When sending JSON data to the server using the HTTP POST, [PUT](https://reqbin.com/Article/HttpPut), or [PATCH](https://reqbin.com/Article/HttpPatch) methods, you must also add the "Content-Type: application/json" header to your request.

Versuche doch mal den header in wpost zu setzen "Content-Type: application/json"

**Comment by Coolj1** [comment]
so jetzt gehts

mit dem Content-Type: application/json gings zwar nicht aber mit application/x-www-form-urlencoded
... die Single Quotes (') am Anfang und Ende hab ich weggelassen👍
danke für deine Hilfe

message$="data={"
message$=message$+chr$(34)+"serial"+chr$(34)+":"+chr$(34)+"112183822934"+chr$(34)
message$=message$+", "+chr$(34)+"limit_type"+chr$(34)+":1"
message$=message$+", "+chr$(34)+"limit_value"+chr$(34)+":50"
message$=message$+"}"

a$ = WPOST$("http://admin:admin0815@192.168.1.228/api/limit/config", message$,0,"application/x-www-form-urlencoded")
wlog a$

Gruß
Hans
